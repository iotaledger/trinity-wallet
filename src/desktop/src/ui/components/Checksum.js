import React from 'react';
import PropTypes from 'prop-types';
import { ADDRESS_LENGTH, VALID_SEED_REGEX } from 'libs/iota/utils';

/**
 * Returns address with highlighted checksum
 */
const Checksum = ({ address }) => {
    const validAddress =
        typeof address === 'string' && address.length === ADDRESS_LENGTH && address.match(VALID_SEED_REGEX);

    return (
        <React.Fragment>
            {!validAddress ? address : address.substr(0, 81)}
            {validAddress && <mark>{address.substr(81)}</mark>}
        </React.Fragment>
    );
};

Checksum.propTypes = {
    /** Target address */
    address: PropTypes.string.isRequired,
};

export default Checksum;
