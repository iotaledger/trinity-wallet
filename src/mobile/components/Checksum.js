import React, { Component } from 'react';
import { View, Text, StyleSheet } from 'react-native';
import PropTypes from 'prop-types';
import { VALID_SEED_REGEX, getChecksum } from 'iota-wallet-shared-modules/libs/iota/utils';
import { width, height } from '../utils/dimensions';
import GENERAL from '../theme/general';

const styles = StyleSheet.create({
    checksum: {
        width: width / 8,
        height: height / 20,
        borderRadius: GENERAL.borderRadiusSmall,
        justifyContent: 'center',
        alignItems: 'center',
    },
    checksumText: {
        fontSize: width / 29.6,
        fontFamily: 'SourceSansPro-Regular',
    },
});

class Checksum extends Component {
    static propTypes = {
        /** Seed value */
        seed: PropTypes.string.isRequired,
        /** Theme settings */
        theme: PropTypes.object.isRequired,
    };

    getChecksumValue() {
        const { seed } = this.props;
        let checksumValue = '...';

        if (seed.length !== 0 && !seed.match(VALID_SEED_REGEX)) {
            checksumValue = '!';
        } else if (seed.length !== 0 && seed.length < 81) {
            checksumValue = '< 81';
        } else if (seed.length === 81 && seed.match(VALID_SEED_REGEX)) {
            checksumValue = getChecksum(seed);
        }

        return checksumValue;
    }

    render() {
        const { theme } = this.props;

        return (
            <View style={[styles.checksum, { backgroundColor: theme.input.bg }]}>
                <Text style={[styles.checksumText, { color: theme.input.color }]}>{this.getChecksumValue()}</Text>
            </View>
        );
    }
}

export default Checksum;
