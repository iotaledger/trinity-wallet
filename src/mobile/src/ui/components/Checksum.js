import size from 'lodash/size';
import React, { Component } from 'react';
import { View, Text, StyleSheet, TouchableOpacity } from 'react-native';
import PropTypes from 'prop-types';
import { MAX_SEED_TRITS, getChecksum } from 'shared-modules/libs/iota/utils';
import { tritsToChars } from 'shared-modules/libs/iota/converter';
import { withTranslation } from 'react-i18next';
import { width } from 'libs/dimensions';
import { Styling } from 'ui/theme/general';
import Icon from 'ui/theme/icons';

const styles = StyleSheet.create({
    checksumContainer: {
        justifyContent: 'center',
        flexDirection: 'row',
        alignItems: 'center',
    },
    checksumText: {
        fontSize: Styling.fontSize3,
        fontFamily: 'SourceSansPro-Regular',
        marginLeft: width / 70,
    },
    checksum: {
        fontSize: Styling.fontSize3,
        fontFamily: 'SourceSansPro-Regular',
        marginLeft: width / 70,
        width: width / 10,
    },
});

export class Checksum extends Component {
    static propTypes = {
        /** @ignore */
        seed: PropTypes.object.isRequired,
        /** @ignore */
        theme: PropTypes.object.isRequired,
        /** @ignore */
        t: PropTypes.func.isRequired,
        /** Show checksum modal */
        showModal: PropTypes.func.isRequired,
    };

    /**
     * Gets the checksum of a seed
     *
     * @method getValue
     *
     * @param {array} input
     *
     * @return {string} Checksum or symbol to be shown
     */
    static getValue(input) {
        const sizeOfInput = size(input);

        if (sizeOfInput !== 0 && sizeOfInput < MAX_SEED_TRITS) {
            return '< 81';
        }

        if (sizeOfInput > MAX_SEED_TRITS) {
            return '> 81';
        } else if (sizeOfInput === MAX_SEED_TRITS) {
            return tritsToChars(getChecksum(input));
        }

        return '...';
    }

    /**
     * Gets the color of the checksum
     *
     * @method getStyle
     *
     * @param {object} theme
     * @param {array} seed
     *
     * @return {object}
     */
    static getStyle(theme, seed) {
        if (size(seed) === MAX_SEED_TRITS) {
            return { color: theme.primary.color };
        }

        return { color: theme.body.color };
    }

    render() {
        const { t, theme, seed } = this.props;

        return (
            <TouchableOpacity
                onPress={() => this.props.showModal()}
                hitSlop={{ top: width / 30, bottom: width / 30, left: width / 30, right: width / 30 }}
            >
                <View style={styles.checksumContainer}>
                    <Icon name="info" size={width / 22} color={theme.body.color} />
                    <Text style={[styles.checksumText, { color: theme.body.color }]}>{t('checksum')}:</Text>
                    <Text style={[styles.checksum, Checksum.getStyle(theme, seed)]}>{Checksum.getValue(seed)}</Text>
                </View>
            </TouchableOpacity>
        );
    }
}

export default withTranslation(['enterSeed', 'global'])(Checksum);
