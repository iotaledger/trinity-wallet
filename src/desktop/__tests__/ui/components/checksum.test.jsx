import React from 'react';
import { shallow } from 'enzyme';

import Checksum from 'ui/components/Checksum';

const props = {
    address: 'ABCDEFGHIJKLMOPRSTUCXYZ9ABCDEFGHIJKLMOPRSTUCXYZ9ABCDEFGHIJKLMOPRSTUCXYZ9ABCDEFGHIJKLMOPRST',
};

describe('Checksum component', () => {
    test('Render the component', () => {
        const wrapper = shallow(<Checksum {...props} />);

        expect(wrapper).toMatchSnapshot();
    });

    test('Address value', () => {
        const wrapper = shallow(<Checksum {...props} />);

        expect(wrapper.text()).toEqual(
            'ABCDEFGHIJKLMOPRSTUCXYZ9ABCDEFGHIJKLMOPRSTUCXYZ9ABCDEFGHIJKLMOPRSTUCXYZ9ABCDEFGHIJKLMOPRST',
        );
    });

    test('Checksum highlight value', () => {
        const wrapper = shallow(<Checksum {...props} />);

        expect(wrapper.find('mark').text()).toEqual('JKLMOPRST');
    });
});
