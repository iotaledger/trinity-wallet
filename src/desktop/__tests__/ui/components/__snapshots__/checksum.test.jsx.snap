// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checksum component Render the component 1`] = `
<Fragment>
  ABCDEFGHIJKLMOPRSTUCXYZ9ABCDEFGHIJKLMOPRSTUCXYZ9ABCDEFGHIJKLMOPRSTUCXYZ9ABCDEFGHI
  <mark>
    JKLMOPRST
  </mark>
</Fragment>
`;
